### PR TITLE
feat(nuxt): custom `history` and `routes` for `app/router.options.ts`

### PR DESCRIPTION
**Pull request by danielroe**
### 🔗 Linked issue

#7122

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

Alongside first-class 'presets' like #7122 and https://github.com/nuxt/framework/issues/4579 which can be configured with a string, it might be useful to allow fully overriding router history + routes for advanced use cases (= escape hatch).

This allows a router options something like this:

```js
import type { RouterConfig } from '@nuxt/schema'
import { createMemoryHistory } from 'vue-router'

export default <RouterConfig> {
  history: process.client ? createMemoryHistory : undefined
}
```

Note that this does not replace https://github.com/nuxt/framework/pull/6980 as hash history has its own challenges with an SSR framework and needs its own implementation.

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by pi0** [comment]
I like the idea of being able to override history and routes via `app/router.options` file.

The factory function for `history` seems good idea too.

Maybe we can also pass default routes to a factory to allow overriding+modifiying routes from config?

(in both regards delaying this PR as well as history for after RC.9 to test better)

**Comment by danielroe** [comment]
Great. Can look into factory routes option as well 👍

Big question in my mind (regardless of the factory function point) is tree-shaking - would be nice to support tree-shaking out auto-generated routes in the event that user fully overrides them, and I'm not sure that rollup will remove them from the build as is. (But then again, nothing stopping user from removing them entirely in the `pages:extend` hook so maybe I'm overthinking this.)

**Comment by pi0** [comment]
Agreed that hook is probably best way to override routes. Do you mind adding function support for routes() in order to merge this?

File: packages/nuxt/src/pages/runtime/router.ts [hunk 1]
```diff
@@ -55,15 +55,16 @@ export default defineNuxtPlugin(async (nuxtApp) => {
   nuxtApp.vueApp.component('NuxtChild', NuxtPage)
 
   const baseURL = useRuntimeConfig().app.baseURL
-  const routerHistory = process.client
-    ? createWebHistory(baseURL)
-    : createMemoryHistory(baseURL)
+  const routerHistory = routerOptions.history?.(baseURL) ??
+    (process.client ? createWebHistory(baseURL) : createMemoryHistory(baseURL))
+
+  const routes = !routerOptions.routes ? _routes : Array.isArray(routerOptions.routes) ? routerOptions.routes : routerOptions.routes(_routes)
```
Comment (pi0) [inline packages/nuxt/src/pages/runtime/router.ts:61]: I vote to accept function for now. We might alternatively use a _merging strategy_ when `routes` as the array is provided...

```suggestion
  const routes = !routerOptions.routes ? _routes : routerOptions.routes(_routes)
```

Comment (danielroe) [reply packages/nuxt/src/pages/runtime/router.ts:61]: oh, I see - you mean support function only 👍

File: packages/nuxt/src/pages/runtime/router.ts [hunk 2]
```diff
@@ -55,15 +55,16 @@ export default defineNuxtPlugin(async (nuxtApp) => {
...
+    (process.client ? createWebHistory(baseURL) : createMemoryHistory(baseURL))
+
+  const routes = !routerOptions.routes ? _routes : routerOptions.routes(_routes)
 
   const initialURL = process.server ? nuxtApp.ssrContext!.url : createCurrentLocation(baseURL, window.location)
   const router = createRouter({
     ...routerOptions,
     history: routerHistory,
-    routes
+    routes: routerOptions.routes || routes
```
Comment (pi0) [inline packages/nuxt/src/pages/runtime/router.ts:67]: Shouldn't be resolved routes from L61?

**Review by pi0 — APPROVED** [review]

**Pull request merged**